File: apps/api/plane/app/serializers/state.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Module imports
from .base import BaseSerializer
from rest_framework import serializers

from plane.db.models import BoardColumn, State, StateGroup


class StateSerializer(BaseSerializer):
    order = serializers.FloatField(required=False)

    class Meta:
        model = State
        fields = [
            "id",
            "project_id",
            "workspace_id",
            "name",
            "color",
            "group",
            "default",
            "description",
            "sequence",
            "order",
            "allow_any_transition",
        ]
        read_only_fields = ["workspace", "project"]

    def validate(self, attrs):
        if attrs.get("group") == StateGroup.TRIAGE.value:
            raise serializers.ValidationError("Cannot create triage state")
        return attrs


class StateLiteSerializer(BaseSerializer):
    class Meta:
        model = State
        fields = ["id", "name", "color", "group"]
        read_only_fields = fields


class BoardColumnSerializer(BaseSerializer):
    state_ids = serializers.SerializerMethodField()

    class Meta:
        model = BoardColumn
        fields = ["id", "project_id", "workspace_id", "name", "sequence", "state_ids"]
        read_only_fields = ["workspace", "project", "state_ids"]

    def get_state_ids(self, obj) -> list:
        return [str(state.id) for state in obj.states.all()]

File: packages/types/src/state.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

export type TStateGroups = "backlog" | "unstarted" | "started" | "completed" | "cancelled";

export interface IState {
  readonly id: string;
  color: string;
  default: boolean;
  description: string;
  group: TStateGroups;
  name: string;
  project_id: string;
  sequence: number;
  workspace_id: string;
  order: number;
  allow_any_transition?: boolean;
}

/** Workflow: allowed transitions keyed by source state id. A state absent from the map allows every target. */
export type TStateTransitionMap = Record<string, string[]>;

export type TStateTransitionPayload = {
  transitions: TStateTransitionMap;
};

/** Board: a column of the project board, owning one or more states. */
export interface TBoardColumn {
  readonly id: string;
  name: string;
  sequence: number;
  project_id: string;
  workspace_id: string;
  state_ids: string[];
}

export type TBoardColumnPayload = {
  name?: string;
  sequence?: number;
  state_ids?: string[];
};

export interface IStateLite {
  color: string;
  group: TStateGroups;
  id: string;
  name: string;
}

export interface IStateResponse {
  [key: string]: IState[];
}

export type TStateOperationsCallbacks = {
  createState: (data: Partial<IState>) => Promise<IState>;
  updateState: (stateId: string, data: Partial<IState>) => Promise<IState | undefined>;
  deleteState: (stateId: string) => Promise<void>;
  moveStatePosition: (stateId: string, data: Partial<IState>) => Promise<void>;
  markStateAsDefault: (stateId: string) => Promise<void>;
};
